onmt/utils/optimizers.py

```diff
@@ -132,119 +153,123 @@ class Optimizer(object):
     rate scheduling beyond what is currently available.
     Also implements necessary methods for training RNNs such
     as grad manipulations.
+    """
 
-    Args:
-      method (:obj:`str`): one of [sgd, adagrad, adadelta, adam]
-      lr (float): learning rate
-      lr_decay (float, optional): learning rate decay multiplier
-      start_decay_steps (int, optional): step to start learning rate decay
-      beta1, beta2 (float, optional): parameters for adam
-      adagrad_accum (float, optional): initialization parameter for adagrad
-      decay_method (str, option): custom decay options
-      warmup_steps (int, option): parameter for `noam` decay
-      model_size (int, option): parameter for `noam` decay
+    def __init__(self,
+                 optimizer,
+                 learning_rate,
+                 learning_rate_decay_fn=None,
+                 max_grad_norm=None):
+        """Initializes the controller.
+
+       Args:
+         optimizer: A ``torch.optim.Optimizer`` instance.
+         learning_rate: The initial learning rate.
+         learning_rate_decay_fn: An optional callable taking the current step
+           as argument and return a learning rate scaling factor.
+         max_grad_norm: Clip gradients to this global norm.
+        """
+        self._optimizer = optimizer
+        self._learning_rate = learning_rate
+        self._learning_rate_decay_fn = learning_rate_decay_fn
+        self._max_grad_norm = max_grad_norm
+        self._training_step = 1
+        self._decay_step = 1
+
+    @staticmethod
+    def from_options(model, opt, checkpoint=None):
```

Comment:
I really like the idea of making objects with something like this instead of the current proliferation of functions whose names start with 'build' or 'make'. Would it make sense, though, for this to be a class method? It is defined on the class `Optimizer` and returns an `Optimizer` instance. Plus, a class method would might make it easier if someone wanted to subclass `Optimizer` later.

Reply:
I think this should be a static function. Otherwise you have to create an instance with an incomplete state.

Reply:
How so?

Reply:
I mean if it's a method you first need an instance, but you haven't yet looked at the options to know how to create it.

Reply:
Class methods (with the decorator `@classmethod`) do not require an instance. The first argument is conventionally `cls` instead of `self` and it is a reference to the class, not an instance of the class. This looks like a good place to use it because the method _does_ use the class. But it does it by hard-coding it in, which would not be necessary with a class method.

Reply:
Oh sorry I was confused by the term "class method" at first. Yes, that makes sense!